src/stores/formula_fingerprints_store.ts

```diff
@@ -0,0 +1,294 @@
+import { ColorGenerator, RangeImpl, reorderZone, setColorAlpha } from "../helpers";
+import { PositionMap } from "../plugins/ui_core_views/cell_evaluation/position_map";
+import {
+  Cell,
+  CellPosition,
+  CellValueType,
+  Color,
+  Command,
+  FormulaCell,
+  isCoreCommand,
+} from "../types/index";
+import { SpreadsheetStore } from "./spreadsheet_store";
+
+/**
+ * This implements formula "fingerprints" to efficient detect formula patterns and anomalies.
+ * A fingerprint is a compact hash representation of a formula's spatial and dependency behavior.
+ *
+ * This is based on the work of Dan Barowy (Williams College), Emery Berger
+ * (UMass Amherst / Microsoft Research), and Benjamin Zorn (Microsoft Research)
+ * as detailed in their paper: "ExceLint: Automatically Finding Spreadsheet Formula Errors"
+ *
+ * Paper: https://dl.acm.org/doi/pdf/10.1145/3276518
+ *
+ * Fingerprints for formulas are computed using *reference vectors*. Each dependency and literal
+ * value gives a reference vector. These vectors collectively represent the formula’s
+ * dependencies and content.
+ *
+ * A reference vector is a 6-dimensional vector. @see ReferenceVector for more details
+ * about the vector's components.
+ *
+ * As an example, the formula `=B1 + Sheet2!B2 + 1` located in A1 produces three reference vectors:
+ * - for `B1`:      (1, 0, 0, 0, 0, 0) (dx=1, dy=0, other components are 0)
+ * - for `Sheet2!`: (2, 0, 1, 0, 0, 0) (dx=2, dy=0, 1 for cross-sheet reference, other components are 0)
+ * - for `1`:       (0, 0, 0, 1, 0, 0) (1 for number, other components are 0)
+ *
+ * The final fingerprint is generated by summing these vectors and hashing the result.
+ * Each fingerprint is then assigned a color to visually represent the formula's behavior.
+ */
+
+/**
+ * A 6-dimensional vector capturing a formula’s dependency or content.
+ */
+type ReferenceVector = {
+  /**
+   * The relative offset in the X coordinate of the referenced cell.
+   * - relative references (B1): coordinates are adjusted based on the formula's position.
+   * - absolute references ($B$1): fixed positions are used, based on spreadsheet coordinates.
+   */
+  dx: number;
+  /**
+   * The relative offset in the Y coordinate of the referenced cell.
+   **/
+  dy: number;
+  /**
+   * Indicates if the reference targets another sheet (0 for same sheet, 1 for another sheet).
+   */
+  crossSheet: number;
+  /**
+   * A literal number (1 if present, 0 otherwise).
+   */
+  number: number;
+  /**
+   * A literal string (1 if present, 0 otherwise).
+   */
+  string: number;
+  /**
+   * A literal boolean (1 if present, 0 otherwise).
+   */
+  boolean: number;
```

Comment:
Normalized formula is debatable. It can help in certain situation, but also hide mistakes in others.
There's no clear answer.

Imagine two different formulas on the same range: `=AVERAGE(A1:A100)` and `=STDEV(A1:A100)`. The current implementation would highlight a mistake such as `=STDEV(A1:A101)`

Actually they mention it as a feature, showing that `=sum(A1:A2)` has the same fingerpring as `=A1+A2` (but it feels like an artificial justification for their implementation TBH).

I'm open to discussion